src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -152,36 +161,60 @@ internal class CompletionHandler : IRequestHandler<CompletionParams, SumType<Com
                     cancellationToken).ConfigureAwait(false);
             }
 
-            if (result.HasValue)
+            if (TryConvertToCompletionList(result, out var completionList))
             {
-                // Set some context on the CompletionItem so the CompletionResolveHandler can handle it accordingly.
-                result = SetResolveData(result.Value, serverKind, request.TextDocument.Uri, projectionResult.Uri);
-
                 if (serverKind == LanguageServerKind.CSharp)
                 {
-                    result = MassageCSharpCompletions(request, documentSnapshot, result.Value);
+                    MassageCSharpCompletions(request, documentSnapshot, completionList);
                 }
+
+                var requestContext = new CompletionRequestContext(documentSnapshot.Uri, projectionResult.Uri, serverKind);
+                var resultId = _completionRequestContextCache.Set(requestContext);
+                SetResolveData(resultId, completionList);
             }
 
-            return result;
+            return completionList;
+
+            static bool TryConvertToCompletionList(SumType<CompletionItem[], CompletionList>? original, out CompletionList completionList)
+            {
+                if (!original.HasValue)
+                {
+                    completionList = null;
+                    return false;
+                }
+
+                if (original.Value.TryGetFirst(out var completionItems))
+                {
+                    completionList = new CompletionList()
+                    {
+                        Items = completionItems,
+                        IsIncomplete = false
+                    };
+                }
+                else if (!original.Value.TryGetSecond(out completionList))
+                {
+                    Debug.Fail("This should be impossible, the completion result should be either a completion list or a set of completion items.");
+                    completionList = null;
+                    return false;
+                }
+
+                return true;
+            }
         }
 
-        private SumType<CompletionItem[], CompletionList>? MassageCSharpCompletions(
+        private void MassageCSharpCompletions(
```

Comment:
Could we get a little doccomment about what "Massage" means in this  context? I see this name and think "This method does unknown magic to change this object somehow." and without deeper context in the area I feel like `GetWordExtent`, `DoNotPreselect` and `RemoveDesignTimeItems` won't clarify a lot.

Reply:
+1. Without context, I'm not sure what "word extent" means here.